forch/faucetizer.py

```diff
@@ -1,49 +1,182 @@
+        """Determine if the switch or port of a device has changed"""
+        if not self.old_learning:
+            return False
+        is_same_switch = self.old_learning.switch != self.new_learning.switch
+        is_same_port = self.old_learning.port != self.new_learning.port
+        return not is_same_switch or not is_same_port
+
+    def is_fulfilled(self):
+        """Determine if the info of device is enough to update output faucet config"""
+        return self.new_learning and self.behavior
+
+    def commit(self):
+        self.old_learning = copy.deepcopy(self.new_learning)
 
 
 class Faucetizer:
     """Collect Faucet information and generate ACLs"""
-    def __init__(self):
-        self._network_state = {}
+    def __init__(self, output_file):
+        self._devices = {}
+        self._base_faucet_config = None
+        self._faucet_config = None
+        self._output_file = output_file
 
     def process_network_state(self, network_state):
-        """Process network state inputs"""
-        self._network_state = proto_dict(network_state, True)
+        """Process network state input"""
+        macs = set()
+        for mac, learning in network_state.device_mac_learnings.items():
+            macs.add(mac)
+            self._process_device_learning(mac, learning)
+        for mac, behavior in network_state.device_mac_behaviors.items():
+            macs.add(mac)
+            self._process_device_behavior(mac, behavior)
+
+        self._faucetize(macs)
+
+    def _process_device_learning(self, mac, learning):
+        device = self._devices.setdefault(mac, Device())
+        device.new_learning = learning
+
+    def _process_device_behavior(self, mac, behavior):
+        device = self._devices.setdefault(mac, Device())
+        device.behavior = behavior
+
+    def process_faucet_config(self, faucet_config):
+        """Process faucet config when base faucet config changes"""
+        self._base_faucet_config = faucet_config
+        self._faucet_config = faucet_config
+        self._faucetize()
+
+    def _faucetize(self, macs=None):
+        if not self._faucet_config:
+            return
+
+        if not macs:
+            macs = self._devices.keys()
+
+        commit = False
+
+        for mac in macs:
+            device = self._devices[mac]
 
-        sorted_rules = list(self._network_state.get("named_match_rules", {}).keys())
-        sorted_rules.sort()
-        sys.stdout.write(f'{sorted_rules}\n')
+            if device.has_migrated():
+                self._reset_port_config(device)
+                commit = True
 
-        sorted_macs = list(self._network_state.get('device_mac_behaviors', {}).keys())
-        sorted_macs.sort()
-        sys.stdout.write(f'{sorted_macs}\n')
+            if device.is_fulfilled():
+                switch = device.new_learning.switch
+                port = device.new_learning.port
+                switch_config = self._faucet_config.get('dps', {}).get(switch, {})
```

Comment:
until it's committed, the _faucet_config maniplated in this function should be a local varaible only.  The class variable should reflect the currently written config.

Reply:
Done. Using a copy of self._out_faucet_config now